processing/src/main/java/io/druid/query/search/SearchQueryRunner.java

```diff
@@ -142,62 +161,125 @@ public SearchQueryRunner(Segment segment)
-            if (timeFilteredBitmap != null) {
-              bitmap = bitmapFactory.intersection(Arrays.asList(timeFilteredBitmap, bitmap));
-            }
-            if (bitmap.size() > 0) {
-              MutableInt counter = new MutableInt(bitmap.size());
-              MutableInt prev = retVal.put(new SearchHit(dimension.getOutputName(), dimVal), counter);
-              if (prev != null) {
-                counter.add(prev.intValue());
-              }
-              if (retVal.size() >= limit) {
-                return makeReturnResult(limit, retVal);
+      switch (strategy) {
+        case AUTO:
+          long totalCard = 0;
+          for (DimensionSpec dimension : dimsToSearch) {
+            final Column column = index.getColumn(dimension.getDimension());
+            if (column != null) {
+              final BitmapIndex bitmapIndex = column.getBitmapIndex();
+              if (bitmapIndex != null) {
+                totalCard += bitmapIndex.getCardinality();
               }
             }
           }
-        }
+
+          // Choose a search query execution strategy depending on the query.
+          // Index-only strategy is selected when
+          // 1) there is no filter,
+          // 2) the total cardinality is very low, or
+          // 3) the filter has a very high selectivity.
+          // The below thresholds are got from some experiments.
+          if (filter == null ||
+              isLowCardinality(bitmapFactory, totalCard) ||
+              isHighSelectivityFilter(index, bitmapFactory, timeFilteredBitmap)) {
+            return indexOnlyExecute(query, searchQuerySpec, timeFilteredBitmap, dimsToSearch, index, limit, bitmapFactory);
+          } else {
+            return cursorBasedExecute(query, searchQuerySpec, filter, dimsToSearch, adapter, limit, descending, interval);
+          }
+
+        case INDEX_ONLY:
+          return indexOnlyExecute(query, searchQuerySpec, timeFilteredBitmap, dimsToSearch, index, limit, bitmapFactory);
+
+        case CURSOR_BASED:
+          return cursorBasedExecute(query, searchQuerySpec, filter, dimsToSearch, adapter, limit, descending, interval);
+
+        default:
+          throw new IAE("Unknown strategy: " + strategy);
       }
 
-      return makeReturnResult(limit, retVal);
+    } else {
+      return cursorBasedExecute(query, searchQuerySpec, filter, dimsToSearch, adapter, limit, descending, interval);
     }
+  }
 
-    final StorageAdapter adapter = segment.asStorageAdapter();
-
-    if (adapter == null) {
-      log.makeAlert("WTF!? Unable to process search query on segment.")
-         .addData("segment", segment.getIdentifier())
-         .addData("query", query).emit();
-      throw new ISE(
-          "Null storage adapter found. Probably trying to issue a query against a segment being memory unmapped."
-      );
+  private static boolean isLowCardinality(final BitmapFactory bitmapFactory, final long totalCard) {
```

Comment:
I disagree, I think this method works better in the SearchQuery since the meaning of "low" and "high" is very subjective and, in this case, determined experimentally for a specific query. It's possible that for a different query, different thresholds would be needed.